packages/flutter/lib/src/widgets/automatic_keep_alive.dart

```diff
@@ -0,0 +1,229 @@
+        childElement.applyWidgetOutOfTurn(build(context));
+      });
+    }
+    return false;
+  }
+
+  VoidCallback _createCallback(Listenable handle) {
+    return () {
+      _handles.remove(handle);
+      if (_handles.isEmpty) {
+        if (SchedulerBinding.instance.schedulerPhase.index < SchedulerPhase.persistentCallbacks.index) {
+          // Build/layout haven't started yet so let's just schedule this for
+          // the next frame.
+          setState(() { _keepingAlive = false; });
+        } else {
+          // We were probably notified by a descendant when they were yanked out
+          // of our subtree somehow. We're probably in the middle of build or
+          // layout, so there's really nothing we can do to clean up this mess
+          // short of just scheduling another build to do the cleanup. This is
+          // very unfortunate, and means (for instance) that garbage collection
+          // of these resources won't happen for another 16ms.
+          //
+          // The problem is there's really no way for us to distinguish these
+          // cases:
+          //
+          //  * We haven't built yet (or missed out chance to build), but
+          //    someone above us notified our descendant and our descendant is
+          //    disconnecting from us. If we could mark ourselves dirty we would
+          //    be able to clean everything this frame. (This is a pretty
+          //    unlikely scenario in practice. Usually things change before
+          //    build/layout, not during build/layout.)
+          //
+          //  * Our child changed, and as our old child went away, it notified
+          //    us. We can't setState, since we _just_ built. We can't apply the
+          //    parent data information to our child because we don't _have_ a
+          //    child at this instant. We really want to be able to change our
+          //    mind about how we built, so we can give the KeepAlive widget a
+          //    new value, but it's too late.
+          //
+          //  * A deep descendant in another build scope just got yanked, and in
+          //    the process notified us. We could apply new parent data
+          //    information, but it may or may not get applied this frame,
+          //    depending on whether said child is in the same layout scope.
+          //
+          //  * We're being notified in the paint phase, or even in a post-frame
+          //    callback. Either way it is far too late for us to make our
+          //    parent lay out again this frame, so the garbage won't get
+          //    collected this frame.
+          //
+          // Long story short, we have to schedule a new frame and request a
+          // frame there, but this is generally a bad practice, and you should
+          // avoid it if possible.
+          scheduleMicrotask(() {
+            setState(() { _keepingAlive = false; });
```

Comment:
Yes. In fact this is wrong in several ways. Good catch, it took me about 3 hours of debugging a weird edge case to figure this out. :-)

The fix is in my next patch: https://github.com/flutter/flutter/pull/11067